### ISSUE TITLE
Error sending request for url (https://repo.hex.pm/packages/gleeunit)

### ISSUE THREAD

**Issue by corani** [gleam-lang/gleam#2708]
This is most likely a local issue, but I haven't been able to get things to work due to some networking issue:

```bash
$ gleam --version
gleam 1.0.0
$ erl -version
Erlang (SMP,ASYNC_THREADS) (BEAM) emulator version 14.2.3

$ gleam new vars
$ cd vars
$ gleam run
  Resolving versions
error: Dependency resolution failed

An error occurred while determining what dependency packages and
versions should be downloaded.
The error from the version resolver library was:

Unable to determine package versions: error sending request for url
(https://repo.hex.pm/packages/gleeunit): error trying to connect: dns
error: failed to lookup address information: Name does not resolve

$ curl https://repo.hex.pm/packages/gleeunit -o out
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1822  100  1822    0     0   3884      0 --:--:-- --:--:-- --:--:--  3876
```

So I can download the file from hex using `curl` just fine, but `gleam run` runs into a DNS error.

(note: both gleam and erlang are installed using ASDF on WSL2/Ubuntu)

**Comment by lpil** [comment]
Thank you

**Comment by corani** [comment]
Yes, consistently. `gleam run` will terminate with the same message in milliseconds each time.

Interestingly, even if I add `repo.hex.pm` to my `hosts` file it'll still fail with the same DNS error.

**Comment by lpil** [comment]
I've just tried adjusting our HTTP client. Tomorrow or later would you be able to download the [nightly](https://github.com/gleam-lang/gleam/releases/tag/nightly) build and see if it works? Thank you

**Comment by lpil** [comment]
Here's the log from another WSL user who got "too many redirects". [client_logs_nc.txt](https://github.com/gleam-lang/gleam/files/14614476/client_logs_nc.txt)

**Comment by corani** [comment]
@lpil I have no clue what's going on. Today after rebooting it suddenly started working. 

I've tried both `1.0.0` and `nightly` a couple of times, and both are fetching from hex without issue now...

Sorry for wasting your time, feel free to close this.

**Comment by lpil** [comment]
Oh dear, it gets even more mysterious... Ok let's close this for now and keep an eye open for any future problems like this. Thank you

**Comment by andho** [comment]
I often run into this problem previously with elixir. And now with gleam as well.

```bash
$ gleam deps update
  Resolving versions
error: HTTP error

A HTTP request failed.
The error from the HTTP client was:

    error sending request for url (https://hex.pm/api/packages/rank/releases/1.0.0)
```

The package the error occurs at always change.

When I changed dns to 8.8.8.8 it worked, but didn't work with the local router or 1.1.1.1.

There is lot's of references to this type of error in elixir/mix as well.

**Comment by andho** [comment]
This just got even weirder. I was testing a local build today and when I did `~/.cargo/bin/gleam deps update` it worked! I didn't realize this though.

Later when I wanted to compare my build with the one from package manager, I found out the same error is still occuring.

So then I downloaded the gleam binary from github releases and still the error persists. But the local build works!

**Comment by lpil** [comment]
Do you have any idea what it might be?

**Comment by andho** [comment]
My suspicion right now is musl libc. I'm assuming if I pull gleam source and build it locally, it will use glibc by default?

**Comment by andho** [comment]
So today I got sometime to test this. Compiled gleam using `cargo build --target=x86_64-unknown-linux-musl --release` and when I try `/home/andho/dev/projects/own/gleam/gleam/target/x86_64-unknown-linux-musl/release/gleam deps download` using this version, I get the same error.

If I run using version compiled with `cargo run` I don't get the error.

And as before, changing DNS to 8.8.8.8 doesn't cause the issue even with musl version.

**Comment by lpil** [comment]
Are you using the precompiled binary for v1.8.0 @sporto? Does it happen every time? Is there anything unusual about your network?

**Comment by sporto** [comment]
@lpil yes using the precompiled binary. It happens quite often, I usually need to run the command several times until eventually it succeeds. I thought the hex server was timing out, but by putting the url in the browser I get an immediate response. I don't think there is anything interesting about my network. I will try to investigate a bit more what is going on.

**Comment by sporto** [comment]
All I figured out so far is that this is network related. Trying to use my wifi this fails constantly. Changing to use a mobile hotspot, this works. Don't know why, the wifi works fine for everything. Any clues what to look for?

**Comment by lpil** [comment]
I don't know enough about networking I'm afraid :(

**Comment by andho** [comment]
@sporto if you are able to compile from source, can you test that out? I found out that glibc version works but musl has issues (the one used in the precompiled binaries). Even with the precompiled binaries, if I change the DNS server it works.

I have commented my findings above. Using DNS 8.8.8.8 works, so that might be why different networks are giving you different results.

**Comment by sporto** [comment]
@andho I got the same result as you

First I tried doing `make build` and copied the binary somewhere to run it.
Still timeouts.

Then I tried doing `cargo run`. 
Then running that to add a new package e.g. `/Users/sebastian/Source/other/gleam/target/debug/gleam add gleesend`
It downloads immediatelly.

**Comment by apsoras** [comment]
I have dealt with similar issues with tls/credentials for other rust-based tools like `uv`, `tealdeer`, etc. For me it boils down to credentials, so `--native-tls`  (see example with `uv` : https://github.com/astral-sh/uv/blob/main/CHANGELOG.md#0110).

I got this error when running `gleam build` and it tried to download the std lib and failed due to credentials. Curl worked fine since it uses system credentials.

### The Fix:

(bump reqwest to 0.13.3)

`reqwest` 0.13 introduced the use of system credentials by default while still using `rust-tls` (https://seanmonstar.com/blog/reqwest-v013-rustls-default/) and I thought that might fix the problem.

I bumped the version for reqwest, updated the feature flag, and gleam compiled and this error is fixed for me when running `gleam build` behind my very strangely proxied network.

Edit: I compiled gleam using stable-gnu and gcc for windows after this change. 

A couple other things changed when running cargo build but that could just be my inexperience with rust/cargo.

Hope this helps, would love to use gleam in more contexts!

```
diff --git a/compiler-cli/Cargo.toml b/compiler-cli/Cargo.toml
index 62cfd6440..3c50d5ede 100644
--- a/compiler-cli/Cargo.toml
+++ b/compiler-cli/Cargo.toml
@@ -22,7 +22,7 @@ tokio = { version = "1", features = ["rt", "rt-multi-thread"] }
 fs_extra = "1"
 tracing-subscriber = { version = "0.3.23", features = ["fmt", "env-filter"] }
 # HTTP client
-reqwest = { version = "0", default-features = false, features = ["rustls-tls"] }
+reqwest = { version = "0.13.3", default-features = false, features = ["rustls"] }
```

**Comment by lpil** [comment]
Brilliant, thank you. Could you make a pull request please 🙏

**Comment by giacomocavalieri** [comment]
I think this can be closed now that #5691 was merged

**Comment by andho** [comment]
I'm not able to reproduce this issue now even using gleam v1.16 from github. So I can't say for sure this fixes the issue (or something else previously did), but closing should be fine. And the PR just updates a lib so not a big issue having it merged.

**Comment by apsoras** [comment]
To be fair, there are a lot of reasons why a http request can fail so this issue likely represents more than one thing!

To reproduce at least the issue I fixed you would need to try using gleam behind a proxy with a self-signed cert, not something that one would do by choice but is common in corporate networks.

I def agree that it likely doesn't have any impact on some of the other problems mentioned above like "too many redirects", etc. which could be non-systemic anyways and more of a Hex thing since people have experienced it with elixir as well.

Edit: Btw I will stop chiming in as I understand this is a maintainer discussion at this point, I'm glad to have helped even in a small way!

**Comment by lpil** [comment]
Thank you everyone!!
